packages/tasit-action/src/TasitAct.js

```diff
@@ -0,0 +1,144 @@
+// shims aren't injected with package import
+import { ethers } from "ethers";
+
+class Utils {
+  static isAddress = address => {
+    return typeof address === "string" && address.match(/^0x[0-9A-Fa-f]{40}$/);
+  };
+
+  static isABI = abi => {
+    return abi && Array.isArray(abi);
+  };
+
+  // https://github.com/ethers-io/ethers.js/blob/db383a3121bb8cf5c80c5488e853101d8c1df353/src.ts/utils/properties.ts#L20
+  static isEthersJsSigner = signer => {
+    return signer && signer._ethersType === "Signer";
+  };
+}
+
+class Subscription {
+  #txPromise;
+  #provider;
+
+  constructor(txPromise, provider) {
+    this.#txPromise = txPromise;
+    this.#provider = provider;
+  }
+
+  on = async (trigger, callback) => {
+    let blockHeight = 0;
+
+    if (trigger === "confirmation") blockHeight = 1;
+    else {
+      throw new Error(`Invalid subscription trigger, use: ["confirmation"]`);
+    }
+
+    if (!callback || typeof callback !== "function") {
+      throw new Error(`Cannot subscribe without a function`);
+    }
+
+    const tx = await this.#txPromise;
+
+    this.#provider.on(tx.hash, async receipt => {
+      const message = {
+        data: {
+          confirmations: receipt.confirmations,
+        },
+      };
+
+      try {
+        await callback(message);
+      } catch (error) {
+        throw new Error(`Callback function with error: ${error.message}`);
+      }
+    });
+  };
+
+  removeAllListeners = async () => {
+    const tx = await this.#txPromise;
+    this.#provider.removeAllListeners(tx.hash);
+  };
+}
+
+export class Contract {
+  address;
+  #provider;
+  #contract;
+
+  constructor(address, abi, wallet) {
+    if (!Utils.isAddress(address) || !Utils.isABI(abi))
+      throw new Error(`Cannot create a Contract without a address and ABI`);
+
+    this.#provider = this.#getDefaultProvider();
+
+    const signerOrProvider = wallet
+      ? wallet.connect(this.#provider)
+      : this.#provider;
+
+    this.#contract = new ethers.Contract(address, abi, signerOrProvider);
+    this.address = this.#contract.address;
+    this.#addFunctionsToContract();
+  }
+
+  // Note: For now, `tasit-account` creates a wallet object
+  setWallet = wallet => {
+    if (!Utils.isEthersJsSigner(wallet))
+      throw new Error(`Cannot set a invalid wallet to a Contract`);
+
+    return new Contract(this.address, this.#contract.interface.abi, wallet);
+  };
+
+  // For tests purpose
+  getProvider = () => {
+    return this.#provider;
+  };
+
+  // Notes:
+  // - JsonRpcProvider will only be used for testing purpose;
+  // - Default provider should be customized (.env file).
+  #getDefaultProvider = () => {
+    const provider = new ethers.providers.JsonRpcProvider();
+    provider.pollingInterval = 50;
+    return provider;
+  };
+
+  #addFunctionsToContract = () => {
+    this.#contract.interface.abi
+      .filter(json => {
+        return json.type === "function";
+      })
+      .forEach(f => {
+        var isWrite = !f.constant;
```

Comment:
`constant` is the older Solidity keyword. Let's also explicitly handle pure and view functions? Or does ethers.js do some magic that buckets those into a constant category?

Reply:
ABI Json `constant` prop is deprecated. I've fixed that using `stateMutability` prop.


> The fields **constant** and payable are deprecated and will be removed in the future. Instead, the **stateMutability** field can be used to determine the same properties.

> stateMutability: a string with one of the following values: pure (specified to not read blockchain state), view (specified to not modify the blockchain state), nonpayable (function does not accept Ether) and payable (function accepts Ether);

Refs: https://solidity.readthedocs.io/en/v0.5.2/abi-spec.html#json

Reply:
@marcelomorgado Perfect, thanks. Will this work in a backwards-compatible way for projects that haven't updated their Solidity version yet? Even if not, I think we can focus on contracts that have switched away from the deprecated prop, at least in the short term.

Reply:
`stateMutability` was introduced v0.4.16 and since v0.5.0 `constant` is deprecated. So I think maybe it won't be a problem.

Reply:
I agree